core/server/src/main/java/alluxio/master/AlluxioMaster.java

```diff
@@ -241,6 +242,14 @@ protected AlluxioMaster(MasterContext masterContext) {
       Configuration.set(Constants.MASTER_RPC_PORT, Integer.toString(mPort));
       mMasterAddress = NetworkAddressUtils.getConnectAddress(ServiceType.MASTER_RPC);
 
+      // Determine port for web server
+      if (Configuration.getInt(Constants.MASTER_WEB_PORT) == 0) {
+        try (ServerSocket serverSocket = new ServerSocket(0)) {
+          int freePort = serverSocket.getLocalPort();
```

Comment:
I've changed the web server to bind to the port in its constructor. Now we can treat it the same way as the rpc and data servers where we bind the port in the constructor so that we know what port it will use, then we start it in the `start()` method.